stl/inc/numeric

```diff
@@ -70,31 +70,41 @@ inline constexpr bool _Plus_on_arithmetic_ranges_reduction_v = false;
 #endif // _STD_VECTORIZE_WITH_FLOAT_CONTROL
 
 template <class _InIt, class _Ty, class _BinOp>
-_NODISCARD _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
+_NODISCARD _CONSTEXPR20_ICE _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
     // return commutative and associative reduction of _Val and [_First, _Last), using _Reduce_op
     _Adl_verify_range(_First, _Last);
     auto _UFirst      = _Get_unwrapped(_First);
     const auto _ULast = _Get_unwrapped(_Last);
-    if constexpr (_Plus_on_arithmetic_ranges_reduction_v<_Unwrapped_t<const _InIt&>, _Ty, _BinOp>) {
-        (void) _Reduce_op; // TRANSITION, VSO-486357
-        return _Reduce_plus_arithmetic_ranges(_UFirst, _ULast, _Val);
-    } else {
+#ifdef __cpp_lib_is_constant_evaluated
+    // TRANSITION, DevCom-878972
+    if (_STD is_constant_evaluated()) {
         for (; _UFirst != _ULast; ++_UFirst) {
             _Val = _Reduce_op(_STD move(_Val), *_UFirst); // Requirement missing from N4713
         }
-
         return _Val;
+    } else
+#endif // __cpp_lib_is_constant_evaluated
+    {
+        if constexpr (_Plus_on_arithmetic_ranges_reduction_v<_Unwrapped_t<const _InIt&>, _Ty, _BinOp>) {
+            (void) _Reduce_op; // TRANSITION, VSO-486357

```

Comment:
Now I am super late to the party, but wouldnt it be possible to move the `is_constant_evaluated` branch in here, so that we get the throughput improvement from the `if constexpr`
```cpp
    if constexpr (_Plus_on_arithmetic_ranges_reduction_v<_Unwrapped_t<const _InIt&>, _Ty, _BinOp>) {
#ifdef __cpp_lib_is_constant_evaluated
        if (_STD is_constant_evaluated()) {
            for (; _UFirst != _ULast; ++_UFirst) {
                _Val = _Reduce_op(_STD move(_Val), *_UFirst); // Requirement missing from N4713
            }
            return _Val;
        }
#endif // __cpp_lib_is_constant_evaluated
        (void) _Reduce_op; // TRANSITION, VSO-486357
        return _Reduce_plus_arithmetic_ranges(_UFirst, _ULast, _Val);
    } else {
        for (; _UFirst != _ULast; ++_UFirst) {
            _Val = _Reduce_op(_STD move(_Val), *_UFirst); // Requirement missing from N4713
        }
        return _Val;
    }
```

Reply:
That still has the loop on both branches of the if constexpr, so I don't think it helps.

Reply:
Good question, I've filed https://github.com/microsoft/STL/issues/414 to keep track of this.